flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java

```diff
@@ -78,13 +75,22 @@ public KvStateServerHandler(
 		final CompletableFuture<KvStateResponse> responseFuture = new CompletableFuture<>();
 
 		try {
-			final InternalKvState<?> kvState = registry.getKvState(request.getKvStateId());
+			final KvStateEntry<?, ?, ?> kvState = registry.getKvState(request.getKvStateId());
 			if (kvState == null) {
 				responseFuture.completeExceptionally(new UnknownKvStateIdException(getServerName(), request.getKvStateId()));
 			} else {
 				byte[] serializedKeyAndNamespace = request.getSerializedKeyAndNamespace();
 
-				byte[] serializedResult = kvState.getSerializedValue(serializedKeyAndNamespace);
+				// here we remove any type check...
+				// Ideally we want to keep that the info match the state.
+				final InternalKvState state = kvState.getState();
+				final KvStateInfo info = kvState.getInfoForCurrentThread();
+
+				byte[] serializedResult = state.getSerializedValue(
```

Comment:
Couldn't we synchronize on `kvState` instead of modifying all `InternalKvState` implementations?

This seems like a much safer alternative than baking in the assumption that `getSerializedValue()` can be called concurrently.

Reply:
It is not an assumption. It is accessed concurrently from the thread pool. Synchronizing on the `kvState` would serialize all accesses, which would slow down (by a lot) performance.

Reply:
Ah, i mistakenly thought that `getSerializedValue()` is also called out of QS.

What would be the overhead of always duplicating serializers within `getSerializedValue()`?

Reply:
"Word on the street" :P is that `Kryo` duplication is pretty expensive. This is why I went for this solution.

Reply:
well we could move the thread-caching into the `InternalKvStates`.

Reply:
True, but after also discussion with Stephan and Aljoscha we concluded that probably it is not the right place, just from a "separation-of-concerns" point of view.

That said, the more I was working on it, the more I was thinking that the `InternalKvStates` could also be a place. But this we can leave for future discussion when we re-prioritize QS.